packages/adblocker/test/engine/engine.test.ts

```diff
@@ -600,6 +600,128 @@ $csp=baz,domain=bar.com
+        'https://foo.com/?utm_source=organic&utm=a',
+      ];
+
+      describe('removes all parameters', () => {
+        let engine: FilterEngine;
+        before(() => {
+          engine = createEngine('||foo.com$removeparam');
+        });
+        for (const url of urls) {
+          it(`removes all params from "${url}"`, () => {
+            const { match, redirect } = engine.match(urlToDocumentRequest(url));
+            expect(match).to.be.true;
+            expect(redirect).not.to.be.undefined;
+            expect(redirect!.body).to.be.eql('');
+            expect(redirect!.contentType).to.be.eql('text/plain');
+            expect(redirect!.dataUrl).to.be.eql('https://foo.com/');
+          });
+        }
+      });
+
+      describe('removes specific parameter', () => {
+        let engine: FilterEngine;
+        before(() => {
+          engine = createEngine('||foo.com$removeparam=utm');
+        });
+        for (const url of urls) {
+          it(`removes "utm" from "${url}"`, () => {
+            const { match, redirect } = engine.match(urlToDocumentRequest(url));
+            expect(match).to.be.true;
+            expect(redirect).not.to.be.undefined;
+            expect(redirect!.body).to.be.eql('');
+            expect(redirect!.contentType).to.be.eql('text/plain');
+            expect(redirect!.dataUrl).not.to.include('?utm=');
+            expect(redirect!.dataUrl).not.to.include('&utm=');
+          });
+        }
+      });
+
+      describe('removes specific parameter regardless of ordering', () => {
+        let engine: FilterEngine;
+        before(() => {
+          engine = createEngine('||foo.com$removeparam=utm');
+        });
+        for (const url of [
+          // First
+          'https://foo.com/?utm=a&utm_source=organic&utm_event=b',
+          // Middle
+          'https://foo.com/?utm_source=organic&utm=a&utm_event=b',
+          // Last
+          'https://foo.com/?utm_source=organic&utm_event=b&utm=a',
+        ]) {
+          it(`removeparam "utm" from "${url}"`, () => {
+            const { match, redirect } = engine.match(urlToDocumentRequest(url));
+            expect(match).to.be.true;
+            expect(redirect).not.to.be.undefined;
+            expect(redirect!.dataUrl).to.be.eql('https://foo.com/?utm_source=organic&utm_event=b');
+          });
+        }
+      });
+
+      it('removes parameter sequentially', () => {
+        const params = ['utm', 'utm_source', 'utm_event'];
+        const engine = createEngine(
+          params.map((param) => `||foo.com$removeparam=${param}`).join('\n'),
+        );
+        let request = urlToDocumentRequest(
+          'https://foo.com/?utm_source=organic&utm_event=b&utm=a',
+        );
+        for (let i = 0; i < params.length; i++) {
+          const { redirect } = engine.match(request);
+          expect(redirect).not.to.be.undefined;
+          request = urlToDocumentRequest(redirect!.dataUrl);
+        }
```

Comment:
It is surprising to me that we need to call the `.match(...)` multiple times to remove all params. Shouldn't a single call to `.match(...)` remove all matching params in a single pass?

Reply:
We didn't know how to return the filters matched since the return type allows only one filter to be returned. That means not the all of filters matched will be sent to the return value of `match` unless we change the shape.

Reply:
Outdated.